### PR TITLE
WeaviateMemoryStore connector

### PR DESCRIPTION
**Pull request by codebrain**
### Description

This adds a WeaviateMemoryStore, which allows for memories to be stored in Weaviate.

I have some outstanding questions, please help!

1. Weaviate has strict naming conventions around class names (i.e. collections). What are the conventions in SK for class names?
2. What are the most appropriate log levels for the connector? Please can you verify these are as expected?
3. I have a suite of integration tests that can run against a Weaviate instance. I have implemented these [integration tests before, in another project](https://github.com/searchpioneer/weaviate-dotnet-client/blob/main/.github/workflows/tests.yml). Any guidance on how these should run for this repo? Ultimately it requires Weaviate running in a docker container.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by codebrain** [comment]
@microsoft-github-policy-service agree

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/Diagnostics/Verify.cs
Comment (dluc) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/Diagnostics/Verify.cs:10]: there are some unused methods which would be safe to remove

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/Http/ApiSchema/GetClassResponse.cs
Comment (dluc) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/Http/ApiSchema/GetClassResponse.cs:5]: this could be internal right?

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/Http/ApiSchema/SearchVectorsResponse.cs
Comment (dluc) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/Http/ApiSchema/SearchVectorsResponse.cs:5]: could you split the file pls, one class per file, matching filename and class name?

**Review by dluc — COMMENTED** [review]
Looking forward to have this ready :-) I think it needs to be rebased and address a few conflicts, than we should find a way to test it against a real instance. Kudos if we include a very quick README in the connector with some details about how to set up the service, and some useful references.

File: dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs
Comment (lemillermicrosoft) [inline dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs:29]: As devis noted, would be nice to have some breadcrumb for folks to set things up to run these tests locally.

File: dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs
Comment (lemillermicrosoft) [inline dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs:14]:
```suggestion
/// Tests for <see cref="s_weaviateMemoryStore" /> collection and upsert operations.
```

uber nit, here and some other documentation comments.

File: dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs
Comment (lemillermicrosoft) [inline dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs:23]: nit: would be nice to add this to the configuration like with other integration tests. Another day, perhaps.

File: samples/apps/copilot-chat-app/webapi/Config/MemoriesStoreOptions.cs
Comment (lemillermicrosoft) [inline samples/apps/copilot-chat-app/webapi/Config/MemoriesStoreOptions.cs:45]:
```suggestion
    /// Gets or sets the configuration for the Weaviate memories store.
```

File: samples/dotnet/kernel-syntax-examples/Program.cs
Comment (lemillermicrosoft) [inline samples/dotnet/kernel-syntax-examples/Program.cs:107]: nit: Example33 and in file name.

**Review by lemillermicrosoft — COMMENTED** [review]
Consider updating `FEATURE_MATRIX.md` to include these changes in the `Connectors and Skill Libraries` section.

File: dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs
Comment (codebrain) [reply dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs:29]: I have added the `docker-compose.yml` alongside these integration tests and updated the comment to instruct the user.

File: dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs
Comment (codebrain) [reply dotnet/src/IntegrationTests/Connectors/Weaviate/WeaviateMemoryStoreTests.cs]: I've removed this requirement and added a docker-compose file to help instantiate a local Weaviate instance.

**Comment by codebrain** [comment]
I have rebased again. Would it be possible to merge this sooner, rather than later, if possible?

Thanks.

**Comment by lemillermicrosoft** [comment]
> I have rebased again. Would it be possible to merge this sooner, rather than later, if possible?
> 
> Thanks.

Thanks for your patience. Will do our best to get this in this week. I can't see the conflicts right now to know what's remaining.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs
Comment (awharrison-28) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs:265]: I would suggest not logging the name of the key as there's a chance it could contain PII

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs
Comment (awharrison-28) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs:272]: See above comment

**Review by awharrison-28 — COMMENTED** [review]
Just a small nit on data being logged

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs
Comment (codebrain) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs:265]: Line 254 of same class, the `string key` is used as the `id` for looking up the record from Weaviate. Is it expected for PII to be in a key? Is there any documentation on what the parameters are expected to look like / contain?

**Comment by codebrain** [comment]
5th rebase (!)

**Comment by lemillermicrosoft** [comment]
> 5th rebase (!)

@codebrain -- It might be worth re-creating the PR with the checkbox of `Enable edits by maintainers` so we can rebase/merge main as appropriate.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/Http/ApiSchema/ObjectResponseResult.cs
Comment (lemillermicrosoft) [inline dotnet/src/Connectors/Connectors.Memory.Weaviate/Http/ApiSchema/ObjectResponseResult.cs:10]: Should this TODO remain?

File: samples/dotnet/kernel-syntax-examples/Example40_Weaviate.cs
Comment (lemillermicrosoft) [inline samples/dotnet/kernel-syntax-examples/Example40_Weaviate.cs:12]: Should add reference in program.cs

**Review by lemillermicrosoft — APPROVED** [review]
Overall LGTM. A few nits, and also suggest re-creating the PR with the checkbox for `Enable edits from Maintainers` to allow us to help with merges/rebases.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs
Comment (awharrison-28) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs:265]: A couple of follow up questions:
- Is the index ID generated service-side? (i.e. what is the batchResponse.Id being returned at line 242?)
- From what I can tell from https://weaviate.io/developers/weaviate/client-libraries/cli, it can be a uuid. Does it have to be a UUID?

Context: 
The key passed to GetAsync and RemoveAsync is expected to be natural language, and could AI-generated. It is possible for this to contain PII. You can see for QdrantMemoryStore, the index ID also has to be a UUID, we had to create separate methods for lookup depending on how the user wants to look up their data. 

*Note that Pinecone allows the index ID to be a provided string of any format.

Comment (codebrain) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs:265]: > Is the index ID generated service-side?

It is provided by the user.

> Does it have to be a UUID?

Yes.

> The key passed to GetAsync and RemoveAsync is expected to be natural language

This is not clear and should be called out in the documentation on the method. I can change the code to accommodate this, but Weaviate expects key to be UUID. If a "natural language" key is provided by the framework the only option I have is to set this as a property on the document and attempt to filter on that value on retrieval. This will not guarantee uniqueness though, which is a concern.

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/Http/ApiSchema/ObjectResponseResult.cs
Comment (codebrain) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/Http/ApiSchema/ObjectResponseResult.cs:10]: Its OK to remain as a JsonObect. I will remove the TODO

File: dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs
Comment (codebrain) [reply dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateMemoryStore.cs:265]: I have omitted the key from the log.

**Pull request closed without merging**
